purple-pill/js/controllers/fontexportController.js

```diff
@@ -32,140 +112,107 @@ function($scope, $http, sharedScope, $timeout) {
+                delete bundleData;
             });
         }
-        
-        function resetProgressBar() {
-            $("#progressbar").animate({"opacity": 0, "width": 0}, /*duration:*/ 0); // (that means "do it immediately!")
-            $("#progresslabel").html("");
-            $("#blob_download").css("display", "none").children("a").unbind("click");
+
+        function exportingGlyphMessage (it, instanceIndex){
+            var msg
+              , currentGlyph = it.value['current_glyph'] + 1 //humans start counting from 1.
+              , totalGlyphs = it.value['total_glyphs']
+              , glyphId = it.value['glyph_id']
+              ;
+            msg = totalInstances + " instances to export."
+            msg += "Calculating glyph '" + glyphId + "' (" + currentGlyph + " of " + totalGlyphs + ")"
+            msg += " of instance #" + (instanceIndex+1);
+            return msg;
+        }
+
+        function calculateGlyphsProgress(it, instanceIndex, totalInstances){
+            var percentage
+              , currentGlyph = it.value['current_glyph']
+              , totalGlyphs = it.value['total_glyphs']
+              ;
+            percentage = (instanceIndex + (currentGlyph / totalGlyphs)) / totalInstances;
+            return percentage;
         }
 
-        function exportFont_compute_glyphs(){
+        var index = 0
+          , totalInstances = exportObjects.length
+          , obj = exportObjects.pop()
+          ;
+        function exportFontComputeGlyphs(){
             var text
-              , current_glyph
-              , total_glyphs
-              , glyph_name
-              , it = generator.next()
+              , percentage
+              , zippedData
+              , it = obj.generator.next()
               ;
             if (!it.done){
-                current_glyph = it.value['current_glyph'];
-                total_glyphs = it.value['total_glyphs'];
-                glyph_id = it.value['glyph_id'];
-                text = total_instances + " instances to export. Calculating glyph '" + glyph_id + "' (" + (current_glyph+1) + " of " + total_glyphs + ") of instance #" + (current_instance+1);
-                setProgress(glyphs_phase_percentage * (current_instance + ((current_glyph+1) / total_glyphs))/total_instances, text);
-                $timeout(exportFont_compute_glyphs, UI_UPDATE_TIMESLICE);
+                text = exportingGlyphMessage(it, index);
+                percentage = calculateGlyphsProgress(it, index, totalInstances);
+                progress.set(percentage, text);
+                $timeout(exportFontComputeGlyphs, UI_UPDATE_TIMESLICE);
             } else {
-                delete generator;
-                if (++current_instance < total_instances) {
-                    generator = getExportGenerator(current_instance);
-                    $timeout(exportFont_compute_glyphs, UI_UPDATE_TIMESLICE);
+                index++;
+                delete obj.generator;
```

Comment:
it would be nice if `ExportObject.prototype` had an interface for this. so that you can do `obj.pruneGenerator()` here.